src/boot/pe.c

```diff
@@ -441,26 +468,89 @@ EFI_STATUS pe_kernel_info(const void *base, uint32_t *ret_compat_address, size_t
         if (!verify_pe(dos, pe, /* allow_compatibility= */ true))
                 return EFI_LOAD_ERROR;
 
+        uint64_t image_base;
+        switch (pe->OptionalHeader.Magic) {
+        case OPTHDR32_MAGIC:
+                image_base = pe->OptionalHeader.ImageBase32;
+                break;
+        case OPTHDR64_MAGIC:
+                image_base = pe->OptionalHeader.ImageBase64;
+                break;
+        default:
+                assert_not_reached();
+        }
+
         /* When allocating we need to also consider the virtual/uninitialized data sections, so parse it out
          * of the SizeOfImage field in the PE header and return it */
-        if (ret_size_in_memory)
-                *ret_size_in_memory = pe->OptionalHeader.SizeOfImage;
+        size_t size_in_memory = pe->OptionalHeader.SizeOfImage;
 
         /* Support for LINUX_INITRD_MEDIA_GUID was added in kernel stub 1.0. */
         if (pe->OptionalHeader.MajorImageVersion < 1)
                 return EFI_UNSUPPORTED;
 
         if (pe->FileHeader.Machine == TARGET_MACHINE_TYPE) {
-                *ret_compat_address = 0;
+                if (ret_entry_point)
+                        *ret_entry_point = pe->OptionalHeader.AddressOfEntryPoint;
+                if (ret_compat_entry_point)
+                        *ret_compat_entry_point = 0;
```

Comment:
NULL rather than 0

Reply:
They are `uint32_t`. Should I change them to `void*`?